extensions/amp-lightbox-gallery/1.0/component.js

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright 2021 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import {Lightbox} from './../../amp-lightbox/1.0/lightbox';
+import {LightboxGalleryContext} from './context';
+import {sequentialIdGenerator} from '../../../src/utils/id-generator';
+import {
+  useContext,
+  useLayoutEffect,
+  useRef,
+  useState,
+} from '../../../src/preact';
+
+const generateLightboxItemKey = sequentialIdGenerator();
+
+/**
+ * @param {!LightboxGalleryDef.Props} props
+ * @return {PreactDef.Renderable}
+ */
+export function LightboxGallery({children}) {
+  const lightboxRef = useRef(null);
+  const lightboxElements = useRef([]);
+  const register = (key, render) => {
+    lightboxElements.current[key] = render();
+  };
+  const deregister = (key) => {
+    delete lightboxElements.current[key];
+  };
+  const context = {
+    deregister,
+    register,
+    open: () => lightboxRef.current.open(),
+  };
+  return (
+    <>
+      <Lightbox ref={lightboxRef} scrollable>
+        {/* TODO: This needs an actual close button UI */}
+        <div
+          role="button"
+          tabindex="0"
+          onClick={() => lightboxRef.current.close()}
+        >
+          Close lightbox
+        </div>
+        <div>{lightboxElements.current}</div>
+      </Lightbox>
+      <LightboxGalleryContext.Provider value={context}>
+        {children}
+      </LightboxGalleryContext.Provider>
+    </>
+  );
+}
+
+/**
+ * @param {!LightboxGalleryDef.WithLightboxProps} props
+ * @return {PreactDef.Renderable}
+ */
+export function WithLightbox({
+  autoLightbox = true,
```

Comment:
Could you pls clarify the intent? If I wrap my component inside the `WithLightbox`, doesn't this already make lightboxable? Why would I specify `autoLightbox=true|false`?

Reply:
"auto" in this case is really about whether it wraps the `children` (in this case, registered elements, so likely individual `img`s) with an outer element that has a click handler which opens the lightbox. This is helpful to do for imgs, and important to disable for carousel, because the default lightbox-wrapping click handler would override clicks to the carousel arrows. I'd totally welcome a better name for this.

Reply:
I can see another use case to be for simple versus complex cases. Simple - `WithLightbox` handles the click fn for you. Complex - you take the `open` fn from the carousel context or we could add an imperative handle here, and create your own interaction patterns for how to trigger opening the lightbox.

Another thing we can do if we don't want to configure this - remove `autoLightbox` entirely and always wrap with a click handler. But consumers who don't want this will have to implement their own interactions, including carousel. I did not go with this approach initially for `BaseCarousel` because it would be rewriting much of the same logic, just not rendering the wrapper `<Comp>`.

Reply:
I see. I think this is more of a "enableActivation" or such.

Reply:
Renamed